zeek/Shaping-Zeek-NDJSON.md

```diff
@@ -0,0 +1,210 @@
+> `v0.30.0` and newer (hence also [Brim](https://github.com/brimdata/brim)
+> `v0.25.0` and newer). If you're looking for docs regarding the legacy
+>`types.json` approach that was used in Zed [`v0.29.0`](https://github.com/brimdata/zed/releases/tag/v0.29.0)
+> (or Brim [`0.24.0`](https://github.com/brimdata/brim/releases/tag/v0.24.0))
+> and older you can find it [here](https://github.com/brimdata/zed/blob/v0.29.0/zeek/README.md).
+
+# Summary
+
+As described in [Reading Zeek Log Formats](Reading-Zeek-Log-Formats.md),
+logs output by Zeek in NDJSON format lose much of their rich data typing that
+was originally present inside Zeek. This detail can be restored using a Zed
+shaper, such as the reference example [`shaper.zed`](shaper.zed)
+that can be found in this directory of the repository.
+
+A full description of all that's possible with shapers is beyond the scope of
+this doc. However, this example for shaping Zeek NDJSON is quite simple and
+is described below.
+
+# Reference Shaper Contents
+
+The example `shaper.zed` may seem large, but ultimately it follows a
+fairly simple pattern that repeats across the many [Zeek log types](https://docs.zeek.org/en/master/script-reference/log-files.html).
+
+## Leading Type Definitions
+
+The top three lines define types that are referenced further below in the main
+portion of the Zed.
+
+```
+type port=uint16;
+type zenum=string;
+type conn_id={orig_h:ip,orig_p:port,resp_h:ip,resp_p:port};
+```
+The `port` and `zenum` types are described further in the [Zed/Zeek Data Type Compatibility](Data-Type-Compatibility.md)
+doc. The `conn_id` type will just save us from having to repeat these fields
+individually in the many Zeek record types that contain an embedded `id`
+record.
+
+## Type Definitions Per Zeek Log `_path`
+
+The bulk of this Zed shaper consists of detailed per-field data type
+definitions for each record in the default set of NDJSON logs output by Zeek.
+These type definitions reference the types we defined above, such as `port`
+and `conn_id`. The syntax for defining primitive and complex types follows the
+relevant sections of the [ZSON Format](../docs/formats/zson.md#3-the-zson-format)
+specification.
+
+```
+...
+type conn={_path:string,ts:time,uid:bstring,id:conn_id,proto:zenum,service:bstring,duration:duration,orig_bytes:uint64,resp_bytes:uint64,conn_state:bstring,local_orig:bool,local_resp:bool,missed_bytes:uint64,history:bstring,orig_pkts:uint64,orig_ip_bytes:uint64,resp_pkts:uint64,resp_ip_bytes:uint64,tunnel_parents:|[bstring]|,_write_ts:time};
+type dce_rpc={_path:string,ts:time,uid:bstring,id:conn_id,rtt:duration,named_pipe:bstring,endpoint:bstring,operation:bstring,_write_ts:time};
+...
+```
+
+> **Note:** See [the role of `_path`](Reading-Zeek-Log-Formats.md#the-role-of-_path)
+> doc for important details if you're using Zeek's built-in [ASCII logger](https://docs.zeek.org/en/current/scripts/base/frameworks/logging/writers/ascii.zeek.html)
```

Comment:
I think this looks odd because "the" is part of the link text but "doc" is not, and it reads well without "doc", so I'd drop it.
```suggestion
> **Note:** See [the role of `_path` ](Reading-Zeek-Log-Formats.md#the-role-of-_path)
> for important details if you're using Zeek's built-in [ASCII logger](https://docs.zeek.org/en/current/scripts/base/frameworks/logging/writers/ascii.zeek.html)
```